interpreter/src/main/java/org/qbicc/interpreter/impl/VmClassLoaderImpl.java

```diff
@@ -0,0 +1,118 @@
+        classes.put(vm.stringClass.getTypeDefinition().getType(), vm.stringClass);
+    }
+
+
+    VmClassLoaderImpl(VmClassImpl clazz, CompilationContext ctxt) {
+        super(clazz);
+        classContext = ctxt.constructClassContext(this);
+    }
+
+    @Override
+    public ClassContext getClassContext() {
+        return classContext;
+    }
+
+    @Override
+    public VmClass loadClass(VmString name) {
+        Assert.checkNotNullParam("name", name);
+        return loadNewClass((VmThreadImpl) Vm.requireCurrentThread(), name);
+    }
+
+    public VmClass defineClass(VmString name, VmArray content, VmObject protectionDomain) {
+        MemoryImpl memory = (MemoryImpl) content.getMemory();
+        ClassFile classFile = ClassFile.of(classContext, ByteBuffer.wrap(memory.getArray()));
+        DefinedTypeDefinition.Builder builder = classContext.newTypeBuilder();
+        classFile.accept(builder);
+        DefinedTypeDefinition defined = builder.build();
+        classContext.defineClass(name.getContent(), defined);
+        LoadedTypeDefinition loaded = defined.load();
+        VmClassImpl.Kind kind;
+        ObjectType type = loaded.getType();
+        VmImpl vm = VmImpl.require();
+        ClassObjectType classLoaderType = (ClassObjectType) vm.classLoaderClass.getObjectType();
+        ClassObjectType throwableType = (ClassObjectType) vm.throwableClass.getObjectType();
+        ClassObjectType threadType = (ClassObjectType) vm.threadClass.getObjectType();
+        if (type.isSubtypeOf(classLoaderType)) {
+            kind = VmClassImpl.Kind.CLASS_LOADER;
+        } else if (type.isSubtypeOf(throwableType)) {
+            kind = VmClassImpl.Kind.THROWABLE;
+        } else if (type.isSubtypeOf(threadType)) {
+            kind = VmClassImpl.Kind.THREAD;
+        } else {
+            kind = VmClassImpl.Kind.OBJECT;
+        }
+        VmClassImpl vmClass = new VmClassImpl(vm.classClass, loaded, null, protectionDomain, kind);
+        classes.put(type, vmClass);
+        return vmClass;
+    }
+
+    VmClassImpl findLoadedClass(ObjectType objectType) {
+        return classes.get(objectType);
+    }
+
+    VmClassImpl findLoadedClass(LoadedTypeDefinition definition) {
+        return findLoadedClass(definition.getType());
+    }
+
+    VmClassImpl loadNewClass(VmThreadImpl thread, VmString intName) {
+        VmClassImpl classLoaderClass = getVmClass();
+        LoadedTypeDefinition clDef = classLoaderClass.getTypeDefinition();
+        ClassTypeDescriptor stringDesc = ClassTypeDescriptor.synthesize(classContext, "java/lang/String");
+        ClassTypeDescriptor classDesc = ClassTypeDescriptor.synthesize(classContext, "java/lang/Class");
+        MethodDescriptor loadClassDesc = MethodDescriptor.synthesize(classContext, classDesc, List.of(stringDesc));
+        return (VmClassImpl) classLoaderClass.getOrCompile(clDef.getMethod(clDef.findMethodIndex("loadClass", loadClassDesc))).invoke(thread, this, intName);
```

Comment:
I think we also need to record the successfully loaded classes in this classloader's `classes` hashtable to ensure we don't attempt to load the class a second time if we've successfully loaded it once

Reply:
Yeah this is a mess right now, and in fact this problem is the reason this is still in draft.  I'm trying out a couple different approaches and we'll see what fits best.